src/dialer/dial-request.js

```diff
@@ -0,0 +1,197 @@
+'use strict'
+
+const AbortController = require('abort-controller')
+const AggregateError = require('aggregate-error')
+const pDefer = require('p-defer')
+const debug = require('debug')
+const log = debug('libp2p:dialer:request')
+log.error = debug('libp2p:dialer:request:error')
+const { AbortError } = require('libp2p-interfaces/src/transport/errors')
+
+const { anySignal } = require('../util')
+const { TokenHolder } = require('./token-holder')
+
+class DialRequest {
+  /**
+   *
+   * @param {object} options
+   * @param {Multiaddr[]} options.addrs
+   * @param {TransportManager} options.transportManager
+   * @param {Dialer} options.dialer
+   */
+  constructor ({
+    addrs,
+    dialAction,
+    dialer
+  }) {
+    this.addrs = addrs
+    this.dialer = dialer
+    this.dialAction = dialAction
+  }
+
+  /**
+   * @async
+   * @param {object} options
+   * @param {AbortSignal} options.signal An AbortController signal
+   * @param {number} options.timeout The max dial time for each request
+   * @returns {Connection}
+   */
+  async run (options) {
+    // Determine how many tokens we need
+    const tokensWanted = Math.min(this.addrs.length, this.dialer.perPeerLimit)
+    // Get the tokens
+    const tokens = this.dialer.getTokens(tokensWanted)
+    // If no tokens are available, throw
+    if (tokens.length < 1) {
+      throw Object.assign(new Error('No dial tokens available'), { code: 'ERR_NO_DIAL_TOKENS' })
+    }
+
+    // For every token, run a multiaddr dial
+    // If there are tokens left, release them
+    // If there are multiaddrs left, wait for tokens to finish
+    const th = new TokenHolder(tokens, this.dialer.releaseToken)
+
+    // Create the dial functions
+    const dials = this.addrs.map(addr => {
+      return () => this._abortableDial(addr, options)
+    })
+
+    const dialResolver = new DialResolver()
```

Comment:
This is quite a lot of complicated code, I wonder if we could essentially remove `TokenHolder` and `DialResolver` and simplify by using `p-fifo` (or something similar)? `p-fifo` is good because you can await on `shift` which resolves when a value gets pushed.

What about something like this?:

```js
const FIFO = require('p-fifo')

const th = new FIFO()
tokens.forEach(t => th.push(t))
const dialAbortControllers = this.addrs.map(() => new AbortController())

try {
  return await Promise.race(this.addrs.map(async (addr, i) => {
    const token = await th.shift() // get token
    let conn
    try {
      const signal = dialAbortControllers[i].signal
      if (signal.aborted) throw new Error('already aborted') // `dialAction` should do this check really...
      conn = await this.dialAction(addr, { ...options, signal: anySignal([signal, options.signal]) })
    } catch (err) {
      th.push(token) // return to token holder on error so another ma can be attempted
      throw err // idk maybe you want to collect these
    }
    return conn
  }))
} finally {
  dialAbortControllers.map(c => c.abort()) // success/failure happened, abort everything else
  tokens.forEach(t => this.dialer.releaseToken(t)) // release tokens back to the dialer
}
```

Reply:
I ended up leveraging this code which got rid of the TokenHolder and DialResolver, so it's now more succinct. I did make a few updates and changes:
1. `pAny` instead of `Promise.race` because we dont want to end after a single failure
2. Tokens are released early. If we have already run through all the addresses when a dial settles, we release it instead of pushing it back onto `p-fifo`. This ensures that if our last dial takes 5 seconds, we don't have to wait for the rest of the tokens to go back into the pool, we will just keep the 1 we're using.